### DISCUSSION TITLE
Sql with parameter is extremely slower than literal compiled

### DISCUSSION THREAD

**Discussion by PaleNeutron** [sqlalchemy/sqlalchemy#9219] [Usage Questions] [answered]
I am using sqlalchemy with Oracle and found some sql is extremely slow.

I finally found that if I compile the query to string with `sql = query.compile(dialect=dialect, compile_kwargs={"literal_binds": True})` and then execute it, the speed is acceptable.

I am wondering why there is significant performance difference between these two ways and if any other sql is slowed down without realizing it.

I can not provide code to reproduce that since I can not provide the database, but I am really wondering which way should I use to execute sql and why sql with parameter is slow.

**Comment by CaselIT** [comment]
Hi,

As mentioned in the docs, you should not use literal_binds with untrusted input. Also not all types are supported.
The main difference here is that the default compile uses parameters, while literal binds inlines them in the sql. 

This is probably something you should report to the driver provider (assuming cx_oracle), since I don't think it's something related with sqlalchemy. Have you tried passing echo=True to the create_engine to see what's the sql that gets executed? Does the same thing happen if you use another tool to execute the sql (like dbeaver or similar?)
The oracle db may also have some configuration regarding prepared statement and what not that may influence this

> I can not provide code to reproduce that since I can not provide the database, but I am really wondering which way should I use to execute sql and why sql with parameter is slow.

If you expect any king of help you should really provide more information, like what kind of sql are you executing? inserts, select, updates? do they have anything in particular? lime how many parameters (ballpark), ctes, subqueries, lots of joins? An example of the sql you find problematic is also useful

**Reply by PaleNeutron** [reply]
The compiled sql  (which takes about 2min) and parameter here is:

```sql
WITH anon_3 AS
  (SELECT main_body_info.id AS id,
          main_body_info.comp_name AS comp_name,
          main_body_info.recent_no_add_tag AS recent_no_add_tag,
          main_body_info.zx_hold_count AS zx_hold_count,
          main_body_info.rating_limit AS rating_limit,
          main_body_info.province AS province,
          main_body_info.city AS city,
          main_body_info.inner_industry_name AS inner_industry_name,
          main_body_info.comp_type AS comp_type,
          main_body_info.name AS name,
          main_body_info.in_library_rule AS in_library_rule,
          main_body_info.main_body_pool_name AS main_body_pool_name,
          main_body_info.rating_pass_time AS rating_pass_time,
          main_body_info.rating_user_name AS rating_user_name,
          main_body_info.total_money AS total_money,
          main_body_info.province_finance_ratio AS province_finance_ratio,
          main_body_info.city_finance_ratio AS city_finance_ratio,
          main_body_info.province_non_finance_ratio AS province_non_finance_ratio,
          main_body_info.city_non_finance_ratio AS city_non_finance_ratio,
          main_body_info.comp_code AS comp_code,
          main_body_info.irm_id AS irm_id,
          main_body_info.bonds_existence AS bonds_existence,
          main_body_info.main_body_pool_id AS main_body_pool_id
   FROM main_body_info
   WHERE main_body_info.name IS NOT NULL
     AND main_body_info.inner_industry_name IS NOT NULL
     AND main_body_info.inner_industry_name IN ('bank')),
     anon_2 AS
  (SELECT amdata.secu_daily_info.dt AS dt,
          anon_3.inner_industry_name AS inner_industry_name,
          anon_3.name AS internal_rating,
          avg(coalesce(amdata.secu_daily_info.chinabond_ytm_ifexe, amdata.secu_daily_info.chinabond_ytm)) AS ytm
   FROM amdata.secu_daily_info
   JOIN amdata.secu_info ON amdata.secu_info.secu_id = amdata.secu_daily_info.secu_id
   JOIN anon_3 ON anon_3.comp_name = amdata.secu_info.issuer
   WHERE anon_3.name IS NOT NULL
     AND amdata.secu_daily_info.dt <= TO_DATE('2023-02-01', 'YYYY-MM-DD')
     AND amdata.secu_daily_info.dt >= TO_DATE('2020-01-01', 'YYYY-MM-DD')
     AND amdata.secu_info.guarantor = 'N/A'
     AND coalesce(amdata.secu_daily_info.chinabond_ytm_ifexe, amdata.secu_daily_info.chinabond_ytm) > 0
     AND amdata.secu_info.issue_type = 'pub'
     AND amdata.secu_info.subordinated = 'Y'
     AND amdata.secu_info.is_pcs = 'N'
   GROUP BY amdata.secu_daily_info.dt,
            anon_3.inner_industry_name,
            anon_3.name),
     anon_1 AS
  (SELECT anon_2.dt AS dt,
          anon_2.inner_industry_name AS inner_industry_name,
          anon_2.internal_rating AS internal_rating,
          anon_2.ytm AS ytm,
          percent_rank() OVER (PARTITION BY anon_2.inner_industry_name,
                                            anon_2.internal_rating
                               ORDER BY anon_2.ytm) AS prank
   FROM anon_2)
SELECT anon_1.dt,
       anon_1.inner_industry_name,
       anon_1.internal_rating,
       anon_1.ytm,
       anon_1.prank
FROM anon_1
WHERE anon_1.dt = TO_DATE('2023-01-31', 'YYYY-MM-DD')
```

And sql with parameter (which takes about 3 hours)  echo  log is:

```
 BEGIN (implicit)

2023-02-02  WITH anon_3 AS 
(SELECT main_body_info.id AS id, main_body_info.comp_name AS comp_name, main_body_info.recent_no_add_tag AS recent_no_add_tag, main_body_info.zx_hold_count AS zx_hold_count, main_body_info.rating_limit AS rating_limit, main_body_info.province AS province, main_body_info.city AS city, main_body_info.inner_industry_name AS inner_industry_name, main_body_info.comp_type AS comp_type, main_body_info.name AS name, main_body_info.in_library_rule AS in_library_rule, main_body_info.main_body_pool_name AS main_body_pool_name, main_body_info.rating_pass_time AS rating_pass_time, main_body_info.rating_user_name AS rating_user_name, main_body_info.total_money AS total_money, main_body_info.province_finance_ratio AS province_finance_ratio, main_body_info.city_finance_ratio AS city_finance_ratio, main_body_info.province_non_finance_ratio AS province_non_finance_ratio, main_body_info.city_non_finance_ratio AS city_non_finance_ratio, main_body_info.comp_code AS comp_code, main_body_info.irm_id AS irm_id, main_body_info.bonds_existence AS bonds_existence, main_body_info.main_body_pool_id AS main_body_pool_id 
FROM main_body_info 
WHERE main_body_info.name IS NOT NULL AND main_body_info.inner_industry_name IS NOT NULL AND main_body_info.inner_industry_name IN (:inner_industry_name_1_1)), 
anon_2 AS 
(SELECT amdata.secu_daily_info.dt AS dt, anon_3.inner_industry_name AS inner_industry_name, anon_3.name AS internal_rating, avg(coalesce(amdata.secu_daily_info.chinabond_ytm_ifexe, amdata.secu_daily_info.chinabond_ytm)) AS ytm 
FROM amdata.secu_daily_info JOIN amdata.secu_info ON amdata.secu_info.secu_id = amdata.secu_daily_info.secu_id JOIN anon_3 ON anon_3.comp_name = amdata.secu_info.issuer 
WHERE anon_3.name IS NOT NULL AND amdata.secu_daily_info.dt <= :dt_1 AND amdata.secu_daily_info.dt >= :dt_2 AND amdata.secu_info.guarantor = :guarantor_1 AND coalesce(amdata.secu_daily_info.chinabond_ytm_ifexe, amdata.secu_daily_info.chinabond_ytm) > :param_1 AND amdata.secu_info.issue_type = :issue_type_1 AND amdata.secu_info.subordinated = :subordinated_1 AND amdata.secu_info.is_pcs = :is_pcs_1 GROUP BY amdata.secu_daily_info.dt, anon_3.inner_industry_name, anon_3.name), 
anon_1 AS 
(SELECT anon_2.dt AS dt, anon_2.inner_industry_name AS inner_industry_name, anon_2.internal_rating AS internal_rating, anon_2.ytm AS ytm, percent_rank() OVER (PARTITION BY anon_2.inner_industry_name, anon_2.internal_rating ORDER BY anon_2.ytm) AS prank 
FROM anon_2)
 SELECT anon_1.dt, anon_1.inner_industry_name, anon_1.internal_rating, anon_1.ytm, anon_1.prank 
FROM anon_1 
WHERE anon_1.dt = :dt_3

 [generated in 0.00071s] {'dt_1': datetime.date(2023, 2, 2), 'dt_2': datetime.date(2020, 1, 1), 'guarantor_1': 'N/A', 'param_1': 0, 'issue_type_1': 'pub', 'subordinated_1': 'Y', 'is_pcs_1': 'N', 'dt_3': datetime.date(2023, 2, 1), 'inner_industry_name_1_1': 'bank'}
```

**Reply by CaselIT** [reply]
I can't see anything strange here. The parameters all have different values, so it shouldn't be a case where the db can see some optimization if the parameters are inline instead of bound.

Just to see if it's something related to sqlalchemy, does executing the parameterized sql with cx_oracle directly has the same performance?

**Comment by zzzeek** [comment, marked as answer]
Oracle has query planning issues that very often can be extremely sensitive to the use of bound parameters.     you'd want to run some [EXPLAIN PLAN](https://docs.oracle.com/cd/B10501_01/server.920/a96533/ex_plan.htm) on your statement .  this is par for the course with Oracle and as the parameters here seem extremely simple (not large binary objects or anything) this is unlikely to be a SQLAlchemy issue, unless the query above you'd prefer if it had a different structure and need to make that work with SQLAlchemy.

**Reply by PaleNeutron** [reply]
Thanks, so what about `literal_binds` compiling all the sql in my project before executing it? Will this solution have some side effects?

**Reply by CaselIT** [reply]
> As mentioned in the docs, you should not use literal_binds with untrusted input. Also not all types are supported.

there is this. If all input is from trusted sources (ie it's not an api where an user can pass something like where parameters ) and the types you use support literal compilation it should work.

Do ensure that the input is trusted though. SQLAlchemy does escaping, but that part has not been security vetted, so you should not blindly trust it

**Reply by zzzeek** [reply]
literal_binds in all cases would not be feasible.  only certain kinds of literal values can be rendered and it will also not work with SQL statement caching at all.   I'd be extremely careful using literal binds.

your best bet is to analyze this query and figure out where it's going wrong, it might be that you need an index somewhere, or something simple like that.

**Reply by CaselIT** [reply]
also since oracle support query hints, sqlalchemy does support these too: 
https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.with_hint
https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.with_statement_hint
